x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/GetGlobalCheckpointsAction.java

```diff
@@ -157,89 +161,178 @@ public TransportAction(
         @Override
         protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
             final ClusterState state = clusterService.state();
-            final Index index = resolver.concreteSingleIndex(state, request);
-            final IndexMetadata indexMetadata = state.getMetadata().index(index);
 
-            if (indexMetadata == null) {
-                // Index not found
-                listener.onFailure(new IndexNotFoundException(request.index));
+            final Index index;
+            try {
+                index = resolver.concreteSingleIndex(state, request);
+            } catch (IndexNotFoundException e) {
+                handleIndexNotReady(request, listener);
                 return;
             }
 
-            final int numberOfShards = indexMetadata.getNumberOfShards();
+            final IndexMetadata indexMetadata = state.getMetadata().index(index);
+            final IndexRoutingTable routingTable = state.routingTable().index(index);
 
-            if (request.waitForAdvance() && numberOfShards != 1) {
-                listener.onFailure(
-                    new ElasticsearchStatusException(
-                        "wait_for_advance only supports indices with one shard. " + "[shard count: " + numberOfShards + "]",
-                        RestStatus.BAD_REQUEST
-                    )
-                );
-                return;
+            if (routingTable.allPrimaryShardsActive()) {
```

Comment:
Should this also check the `waitForAdvance` flag (or another `waitForYellow` flag) to still allow a get checkpoints that just returns the available checkpoints regardless of primary failures. I think we should strive to return current state immediately if there are no `wait_for_xxx` options enabled on the request.